### DISCUSSION TITLE
Vuex action(s) in beforeRouteEnter vue-router navigation guard [using vuex-smart-module]

### DISCUSSION THREAD

**Discussion by jagged3dge** [quasarframework/quasar#11930] [General - Components / Directives / etc] [answered]
The option of using `vuex-smart-module` for better Typescript support is just brilliant!
Thank you Quasar maintainers for this gem!

With that said, I'm having a slight issue with accessing the store in my root component's `beforeRouteEnter` navigation guard.

The purpose of the usage is to determine if a user is authenticated. If yes, then continue to the `/home` route. Otherwise, redirect to `/login`. Fairly simple. 

```ts
// pages/Index.vue

import { defineComponent } from 'vue'
import { useStore } from '../store'

export default defineComponent({
  name: 'PageIndex',

  beforeRouteEnter: async (to, from, next) => {
    if (!!to.meta.public) return next()

    const store = useStore()
    // console.log('store =', store)

    const hasAuth = tokenUtils.getAccessToken()
    // TODO: Replace ^^^ with auth store getter

    if (!hasAuth) return next({ name: 'Login' })

    if (store.modules.auth.state.user) return next({ name: 'Home' })
    // ^^^ TypeError: Cannot read properties of undefined (reading 'state')

    try {
      await store.modules.auth.actions.getProfile()
    } catch (err) {
      if (process.env.DEV) console.log('err =', err)
      return next({ name: 'Login' })
    }
  },

  setup() {
    // other code
  },
}
```

But when I run this, I get the following warning/error messages:

```
WARN: [Vue warn]: inject() can only be used inside setup() or functional components.     [runtime-core.esm-bundler.js?d976:6620]

ERROR: TypeError: Cannot read properties of undefined (reading 'state')     [vue-router.esm-bundler.js?180b:3295]
```

What is the right way to access the `vuex-smart-store` store in a component's `beforeRouteEnter` guard?

**Comment by TobyMosque** [comment, marked as answer]
composables only works in the setup.
in your particular case, preFetch will fit like a glove.:
https://quasar.dev/quasar-cli/prefetch-feature

```js
import { defineComponent } from 'vue'
import { preFetch } from 'quasar/wrappers'

export default defineComponent({
  name: 'PageIndex',

  preFetch: preFetch(async ({ store, currentRoute: to, previousRoute: from, redirect: next }) => {
    if (!!to.meta.public) return next()
    // console.log('store =', store)

    const hasAuth = tokenUtils.getAccessToken()
    // TODO: Replace ^^^ with auth store getter

    if (!hasAuth) return next({ name: 'Login' })

    if (store.modules.auth.state.user) return next({ name: 'Home' })
    // ^^^ TypeError: Cannot read properties of undefined (reading 'state')

    try {
      await store.modules.auth.actions.getProfile()
    } catch (err) {
      if (process.env.DEV) console.log('err =', err)
      return next({ name: 'Login' })
    }
  }),

  setup() {
    // other code
  },
}
```

In this particular case, I would prefer to transform the auth logic in a composable in move that to the setup.
yep, the setup can be async, as long you can keep the suspense:
https://v3.vuejs.org/guide/migration/suspense.html

IMHO: u would stop to waste time with Vuex 4 and move to Pinia.

**Comment by jagged3dge** [comment]
Thank you @TobyMosque for pointing me to the `prefetch` feature of Quasar. After reading the docs for `prefetch`, you are absolutely correct that it does fit my purpose exactly.

I shall strongly consider switching to Pinia.

For clarity's sake, if I wasn't using Quasar CLI (`prefetch` only works in projects using Quasar CLI), I must wrap the logic as a composable and use an `async setup()` function with [suspense ](https://v3.vuejs.org/guide/migration/suspense.html).

Did I get that correctly?
